File: flutter_frontend/lib/assets/youtube/requests/youtubeStreamInfoRequest.dart
import 'dart:convert';
import 'package:flutter/material.dart';
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'package:http/http.dart' as http;
import 'package:flutter_frontend/static/constants.dart';
import 'package:flutter_frontend/pages/management/editYoutubeVideoPage.dart';

class YoutubeStreamInfoRequest extends StatefulWidget {
  const YoutubeStreamInfoRequest({Key? key}) : super(key: key);

  @override
  State<YoutubeStreamInfoRequest> createState() =>
      _YoutubeStreamInfoRequestState();
}

class _YoutubeStreamInfoRequestState extends State<YoutubeStreamInfoRequest> {
  final FlutterSecureStorage _storage = const FlutterSecureStorage();

  bool _loading = true;
  String? _error;
  Map<String, dynamic>? _streamInfo;

  @override
  void initState() {
    super.initState();
    _fetchStreamInfo();
  }

  Future<void> _fetchStreamInfo() async {
    setState(() {
      _loading = true;
      _error = null;
    });

    try {
      final token = await _storage.read(key: 'token');
      final youtubeSerial = await _storage.read(key: 'youtubeSerial');

      if (token == null || youtubeSerial == null) {
        setState(() {
          _error = 'Missing authentication token or YouTube serial.';
          _loading = false;
        });
        return;
      }

      final response = await http.get(
        Uri.parse('$server/get/youtube_stream_data/$youtubeSerial/'),
        headers: {'Authorization': token, 'Content-Type': 'application/json'},
      );

      if (response.statusCode != 200) {
        final errorData = json.decode(response.body);
        throw Exception(errorData['message'] ?? 'Failed to fetch stream info.');
      }

      final jsonData = json.decode(response.body);

      setState(() {
        _streamInfo = jsonData['data'];
        _loading = false;
      });
    } catch (e) {
      setState(() {
        _error = e.toString();
        _loading = false;
      });
    }
  }

  Future<void> _onEditPressed() async {
    await _storage.write(
      key: 'editYoutubeVideoSerial',
      value: _streamInfo!['serial'].toString(),
    );

    if (!mounted) return;
    Navigator.of(
      context,
    ).push(MaterialPageRoute(builder: (_) => EditYoutubeVideoPage()));
  }

  @override
  Widget build(BuildContext context) {
    if (_loading) {
      return const Scaffold(body: Center(child: CircularProgressIndicator()));
    }

    if (_error != null) {
      return Scaffold(
        body: Center(
          child: Text(
            _error!,
            style: const TextStyle(color: Colors.red),
            textAlign: TextAlign.center,
          ),
        ),
      );
    }

    return Scaffold(
      body: SingleChildScrollView(
        padding: const EdgeInsets.all(24.0),
        child: Center(
          child: ConstrainedBox(
            constraints: const BoxConstraints(maxWidth: 800),
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.center,
              children: [
                Text(
                  _streamInfo!['title'],
                  style: Theme.of(context).textTheme.headlineMedium,
                  textAlign: TextAlign.center,
                ),
                const SizedBox(height: 16),
                Text(_streamInfo!['description'], textAlign: TextAlign.center),
                const SizedBox(height: 32),

                ElevatedButton.icon(
                  onPressed: _onEditPressed,
                  icon: const Icon(Icons.edit),
                  label: const Text('Edit Video'),
                  style: ElevatedButton.styleFrom(
                    padding: const EdgeInsets.symmetric(
                      horizontal: 32,
                      vertical: 16,
                    ),
                  ),
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }
}

File: frontend/src/Components/Requests/YoutubeStreamInfoRequest.js
import React, { useEffect, useState } from "react";
import server from "../Static/Constants";
import { useLocation } from "react-router-dom";

const YoutubeStreamInfoRequest = () => {
    const location = useLocation();
    const search = location.search;

    const [streamInfo, setStreamInfo] = useState(null);
    const [loading, setLoading] = useState(true);
    const [error, setError] = useState("");

    useEffect(() => {
        const fetchStreamInfo = async () => {
            const params = new URLSearchParams(search);
            const streamSerial = params.get("serial");

            if (!streamSerial) {
                setError("No stream serial provided in the URL.");
                setLoading(false);
                return;
            }

            try {
                const response = await fetch(
                    `${server}/get/youtube_stream_data/${streamSerial}/`,
                    {
                        method: "GET",
                        headers: {
                            Authorization: localStorage.getItem("token"),
                            "Content-Type": "application/json",
                        },
                    }
                );

                if (!response.ok) {
                    const errorData = await response.json();
                    throw new Error(errorData.message || "Failed to fetch stream info.");
                }

                const json = await response.json();
                setStreamInfo(json.data);
            } catch (err) {
                setError(err.message);
            } finally {
                setLoading(false);
            }
        };

        fetchStreamInfo();
    }, [search]);

    if (loading) return <p>Loading...</p>;
    if (error) return <p style={{ color: "red" }}>{error}</p>;

    return (
        <div>
            <h1>{streamInfo.title}</h1>
            <p>{streamInfo.description}</p>
        </div>
    );
};

export default YoutubeStreamInfoRequest;
